### PR TITLE
Qualcomm AI Engine Direct - Enable AR-N model for prompt processing in hybrid mode

### PR DESCRIPTION
**Pull request by shewu-quic**
## Summary:

- Add `--max_seq_len` to refer to maximum number of tokens that the model can process & consider at once to generate predictions/responses.
- Add `--prefill_ar_n` to determine the number of tokens to consume and the number of logits to produce for prompt processor in hybrid mode.
- Remove prefill mode
- The best AR-N depends on model size, use case, HW capabilities, etc.
  - Suggest to profile on-target latency measurements and decide best config
 
## Test Plan
- Try to find the best AR-N with CL=2048 in hybrid mode
  - Based on the below table, the best AR-N is 256
image

- Ensure accuracy for Stories Llama 16a4w, Prompt: "Once upon a time"

  - Before(Current Mainline) KV mode
```
Once upon a time, there was a little girl named Lily. She loved to play with her toys and her favorite toy was a big, red ball. One day, Lily's mom asked her to help her with the laundry. Lily was happy to help and she put all the clothes in the washing machine. After the clothes were washed, Lily's mom asked her to help her hang them up to dry. Lily saw a big, black sheet hanging on the line and she wanted to help. She grabbed the sheet and tried to hang it up
```
  - After(This PR) Shift Pointer and prefill_ar_len = 8 and max_seq_len=128 in hybrid mode 
```
I 00:00:00.377477 executorch:runner.cpp:346] Prompt Processor: total 5 tokens (AR-8 * 1 iters)
I 00:00:00.927398 executorch:runner.cpp:446] 	Prompt Tokens: 5    Generated Tokens: 122
I 00:00:00.927487 executorch:runner.cpp:452] 	Model Load Time:		0.375000 (seconds)
I 00:00:00.927502 executorch:runner.cpp:462] 	Total inference time:		0.550000 (seconds)		 Rate: 	221.818182 (tokens/second)
I 00:00:00.927513 executorch:runner.cpp:470] 		Prompt evaluation:	0.008000 (seconds)		 Rate: 	625.000000 (tokens/second)
I 00:00:00.927522 executorch:runner.cpp:481] 		Generated 122 tokens:	0.542000 (seconds)		 Rate: 	225.092251 (tokens/second)
I 00:00:00.927530 executorch:runner.cpp:489] 	Time to first generated token:	0.008000 (seconds)
I 00:00:00.927538 executorch:runner.cpp:496] 	Sampling time over 122 tokens:	0.025000 (seconds)

INFO:root:Results[0]:
Once upon a time, there was a little girl named Lily. She loved to play with her toys and her favorite toy was a big, red ball. One day, Lily's mom asked her to help her with the laundry. Lily was happy to help and she put all the clothes in the washing machine. 
After the clothes were washed, Lily's mom asked her to help her hang them up to dry. Lily saw a big, black iron on the counter and asked her mom what it was for. Her mom explained that it was used to make clothes look
```
  - After(This PR) Smart Mask and prefill_ar_len = 8 and max_seq_len=128 in hybrid mode 
```
I 00:00:00.928392 executorch:runner.cpp:446] 	Prompt Tokens: 5    Generated Tokens: 122
I 00:00:00.928457 executorch:runner.cpp:452] 	Model Load Time:		0.367000 (seconds)
I 00:00:00.928473 executorch:runner.cpp:462] 	Total inference time:		0.559000 (seconds)		 Rate: 	218.246869 (tokens/second)
I 00:00:00.928484 executorch:runner.cpp:470] 		Prompt evaluation:	0.039000 (seconds)		 Rate: 	128.205128 (tokens/second)
I 00:00:00.928493 executorch:runner.cpp:481] 		Generated 122 tokens:	0.520000 (seconds)		 Rate: 	234.615385 (tokens/second)
I 00:00:00.928501 executorch:runner.cpp:489] 	Time to first generated token:	0.039000 (seconds)
I 00:00:00.928509 executorch:runner.cpp:496] 	Sampling time over 122 tokens:	0.035000 (seconds)

INFO:root:Results[0]:
Once upon a time, there was a little girl named Lily. She loved to play with her toys and her favorite toy was a big, red ball. One day, Lily's mom asked her to help her with the laundry. Lily was happy to help and she put all the clothes in the washing machine. 
After the clothes were washed, Lily's mom asked her to help her hang them up to dry. Lily saw a big, black iron on the counter and asked her mom what it was for. Her mom explained that it was used to make clothes smooth
```

## Command
```
python examples/qualcomm/oss_scripts/llama/llama.py -b build-android --checkpoint ~/.llama/checkpoints/Llama3.2-1B-Instruct/consolidated.00.pth --params ~/.llama/checkpoints/Llama3.2-1B-Instruct/params.json --tokenizer_model ~/.llama/checkpoints/Llama3.2-1B-Instruct/tokenizer.model --prompt $'<|start_header_id|>system<|end_header_id|>\n\nYou are a funny chatbot.<|eot_id|><|start_header_id|>user<|end_header_id|>\n\nCould you tell me about Facebook?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\n' --temperature 0 --llama_model llama3_2 --model_mode hybrid --ptq 16a4w -m SM8650 -H ${HOST} -s ${DEVICE}-a ${ARTIFACTS} --max_seq_len 2048 --prefill_ar_len 256 --num_sharding 4 --kv_updater shift_pointer
```

### CODE REVIEW

**Comment by shewu-quic** [comment]
Hi @cccclai, @billmguo,

This PR enables the AR-N model for prompt processing in hybrid mode.
Could you please help to take a look?

Regarding the change in lower_module_backend.py, it is intended to prevent a double deletion of the persistent buffer. I observed that the buffer (freq_cos and freq_sin) is copied to each delegate node (due to graph sharding), and the original buffer is eventually deleted. Since each copied buffer shares the same target, this would result in a double deletion.

**Comment by cccclai** [comment]
Awesome! Thank you for getting it to work so quickly. Can you help fix these errors?
```
executorch/examples/qualcomm/oss_scripts/llama/runner/io_manager.cpp:574:7: error: unused variable 'ptr' [-Werror,-Wunused-variable]
  574 |   IO* ptr = static_cast<IO*>(data_ptr_.get());
      |       ^~~
executorch/examples/qualcomm/oss_scripts/llama/runner/io_manager.cpp:1089:11: error: unused variable 'cache_len' [-Werror,-Wunused-variable]
 1089 |   int32_t cache_len = methods_meta[0]->input_tensor_meta(0)->sizes()[1];
      |           ^~~~~~~~~
executorch/examples/qualcomm/oss_scripts/llama/runner/io_manager.cpp:1306:7: error: unused variable 'ptr' [-Werror,-Wunused-variable]
 1306 |   IO* ptr = static_cast<IO*>(data_ptr_.get());
```

File: examples/qualcomm/oss_scripts/llama/README.md [hunk 1]
```diff
@@ -8,11 +8,10 @@ This file provides you the instructions to run LLAMA model with different parame
...
-
 KV Cache Mode: In KV Cache mode, the model takes in a single previous token and generates the next predicted token along with its KV cache. It is efficient for generating subsequent tokens after the initial prompt.
 
-Hybrid Mode: Hybrid mode leverages the strengths of both batch prefill and KV cache modes to optimize token generation speed. Initially, it uses prefill mode to efficiently generate the prompt's key-value (KV) cache. Then, the mode switches to KV cache mode, which excels at generating subsequent tokens.
+Hybrid Mode: Hybrid mode leverages the strengths of both AR-N model and KV cache modes to optimize token generation speed. Initially, it uses AR-N model to efficiently generate the prompt's key-value (KV) cache. Then, the mode switches to KV cache mode, which excels at generating subsequent tokens.
+  - AR-N model: The auto-regression (AR) length determines the number of tokens to consume and the number of logits to produce. Use it to process the prompt and generate the key-value (kv) cache, which serves as a prompt processor in hybrid mode.
```
Comment (cccclai) [inline examples/qualcomm/oss_scripts/llama/README.md:14]: Can we add the diagram you shared as part of the readme? It's much easier to understand with it.

Comment (shewu-quic) [reply examples/qualcomm/oss_scripts/llama/README.md:14]: Certainly, that will not be a problem.

**Review by cccclai — APPROVED** [review]
Thanks! And the lint

**Comment by shewu-quic** [comment]
> Thanks! And the lint

Woops, Thanks for your effort.

**Comment by cccclai** [comment]
Hey it seems like some merge conflict, can you rebase?

File: examples/qualcomm/oss_scripts/llama/runner/io_manager.h [hunk 2]
```diff
@@ -287,6 +308,9 @@ class SmartMaskIoMgr : public IoMgrBase {
   std::string prefill_forward_name_;
   std::string kv_forward_name_;
   const bool use_int64_token_{false};
+  // If the cache length is zero, it indicates a BERT model, which does not use
+  // position ids or KV cache inputs.
+  const bool is_bert_{false};
```
Comment (swolchok) [inline examples/qualcomm/oss_scripts/llama/runner/io_manager.h:313]: why waste a byte storing this separately rather than a private method like so?

```
bool is_bert() const {
  return prefill_cache_len_ == 0;
}
```
